src/utilities/to_crd.py

```diff
@@ -0,0 +1,423 @@
+                                "image": image,
+                                "ports": {},
+                                "command": self.command,
+                                "resource": self.resource.to_crd()
+                                }
+                                ]
+                            }
+                        }
+                    }
+                }
+
+        return result
+
+
+def parse_tasks(obj, parameters, prerequisites):
+    tasks = {}
+
+    if len(obj.get("tasks", [])) == 0:
+        raise RuntimeError("should specify at least one task")
+
+    for task in obj["tasks"]:
+        task = Task.parse(task, parameters, prerequisites)
+        if task.role in tasks:
+            raise RuntimeError("role name in task should be unique")
+        tasks[task.role] = task
+
+    return tasks
+
+
+@attr.s
+class Job(object):
+    name = attr.ib() # required
+    version = attr.ib() # optional
+    contributor = attr.ib() # optional
+    desc = attr.ib() # optional
+    virtual_cluster = attr.ib() # optional default to "default"
+    retry_count = attr.ib() # optional default to 0
+    gpu_type = attr.ib() # optional
+    tasks = attr.ib() # required
+
+    @staticmethod
+    def parse(obj):
+        protocol_version = obj.get("protocol_version", "v2")
+        if protocol_version != "v2":
+            raise RuntimeError("unknown protocol version " + version)
+
+        name = get_required_field(obj, "name", "name field is required")
+
+        type = obj.get("type")
+        if type != "job":
+            raise RuntimeError("type must not be omitted and should be job")
+
+        version = obj.get("version")
+        contributor = obj.get("contributor")
+        desc = obj.get("description")
+        vc = obj.get("virtualCluster")
+        retry_count = int(obj.get("retryCount", "0"))
+        gpu_type = obj.get("gpuType")
+
+        parameters = parse_parameters(obj.get("parameters"), {})
+        prerequisites = parse_prerequisites(obj.get("prerequisites"))
+
+        tasks = parse_tasks(obj, parameters, prerequisites)
+
+        return Job(name=name, version=version, contributor=contributor, desc=desc,
+                virtual_cluster=vc, retry_count=retry_count, gpu_type=gpu_type, tasks=tasks)
+
+
+    def process_prerequisite(self):
+        for role, task in self.tasks.items():
+            task.process_prerequisite()
+
+    def validate(self):
+        if self.name is not None and re.match("^[A-Za-z0-9._~]+$", self.name) is None:
+            raise RuntimeError("job name should match [A-Za-z0-9._~]+")
+
+        for role, task in self.tasks.items():
+            task.validate()
+
+
+    def to_crd(self):
+        desc = self.desc
+        if desc is None:
+            desc = "job of " + self.name
+
+        retry_policy = {"fancyRetryPolicy": False}
+        restart_policy = "Never"
+
+        result = {
+                "apiVersion": "launcher.microsoft.com/v1",
```

Comment:
API is changed:
apiVersion: frameworkcontroller.microsoft.com/v1
, please refer https://github.com/Microsoft/frameworkcontroller/tree/master/example/framework

I will update int bed controller now